ziti/src/main/kotlin/org/openziti/api/Authentication.kt

```diff
@@ -0,0 +1,299 @@
+import org.openziti.util.ZitiLog
+import java.net.URI
+import java.net.URLEncoder
+import java.net.http.HttpClient
+import java.net.http.HttpRequest
+import java.net.http.HttpResponse
+import java.nio.charset.StandardCharsets
+import java.security.MessageDigest
+import java.time.OffsetDateTime
+import java.util.Base64
+import java.util.function.Consumer
+import javax.net.ssl.SSLContext
+import kotlin.random.Random
+
+
+interface ZitiAuthenticator {
+    enum class TokenType {
+        BEARER, API_SESSION
+    }
+
+    data class ZitiAccessToken(
+        val type: TokenType,
+        val token: String,
+        val expiration: OffsetDateTime
+    )
+    suspend fun login(): ZitiAccessToken
+    suspend fun refresh(): ZitiAccessToken
+}
+
+internal fun authenticator(ep: String, ssl: SSLContext, oidc: Boolean): ZitiAuthenticator =
+    if (oidc)
+        InternalOIDC(ep, ssl)
+    else
+        LegacyAuth(ep, ssl)
+
+class LegacyAuth(val ep: String, val ssl: SSLContext) : ZitiAuthenticator, Logged by ZitiLog() {
+
+    private val auth = Authenticate().apply {
+        val info = SystemInfoProvider().getSystemInfo()
+        sdkInfo = SdkInfo()
+            .type("ziti-sdk-java")
+            .version(Version.version)
+            .branch(Version.branch)
+            .revision(Version.revision)
+            .appId(ZitiImpl.appId)
+            .appVersion(ZitiImpl.appVersion)
+        envInfo = EnvInfo()
+            .arch(info.arch)
+            .os(info.os)
+            .osRelease(info.osRelease)
+            .osVersion(info.osVersion)
+        configTypes = listOf("all")
+    }
+
+    private val http = HttpClient.newBuilder()
+        .sslContext(ssl)
+        .executor(Dispatchers.IO.asExecutor())
+
+    private val api = ApiClient().apply {
+        setHttpClientBuilder(http)
+        updateBaseUri(ep)
+    }
+
+    override suspend fun login(): ZitiAuthenticator.ZitiAccessToken {
+        val authApi = AuthenticationApi(api)
+        val session = authApi.authenticate("cert", auth).await()
+        api.requestInterceptor = Consumer {
+            req -> req.header("zt-session", session.data.token)
+        }
+        return ZitiAuthenticator.ZitiAccessToken(
+            ZitiAuthenticator.TokenType.API_SESSION,
+            session.data.token,
+            session.data.expiresAt
+        )
+    }
+
+    override suspend fun refresh(): ZitiAuthenticator.ZitiAccessToken {
+        val currentApiSessionApi = CurrentApiSessionApi(api)
+        val session = currentApiSessionApi.currentAPISession.await()
+        return ZitiAuthenticator.ZitiAccessToken(
+            ZitiAuthenticator.TokenType.API_SESSION,
+            session.data.token,
+            session.data.expiresAt
+        )
+    }
+}
+
+class InternalOIDC(val ep: String, ssl: SSLContext): ZitiAuthenticator, Logged by ZitiLog() {
+
+    companion object {
+        const val CLIENT_ID = "openziti"
+        const val internalRedirect = "http://localhost:8080/auth/callback"
```

Comment:
i assume this is just hardcoded by the controller or there's no redirect uri matching?

Reply:
I believe this is hardcoded in controller